File: libraries/typescript/src/engines/gemini.ts
/**
 * Gemini Live engine — marker class for Patter client dispatch.
 *
 * Selects Google's Gemini Live native-audio API (audio-in → audio-out,
 * emotion-aware). Separate marker from the OpenAI / ElevenLabs engines
 * because the client dispatches to {@link import('../providers/gemini-live').GeminiLiveAdapter}
 * when this marker is passed to ``Patter.agent({ engine })``.
 *
 * Like the other engine markers this is a tiny immutable config object: it
 * carries credentials / voice / model only. The runtime session lives in
 * ``GeminiLiveAdapter`` (constructed server-side by ``buildAIAdapter``).
 *
 * @example
 * ```ts
 * import { Patter, Twilio, GeminiLive } from "getpatter";
 *
 * const phone = new Patter({ carrier: new Twilio(), phoneNumber: "+1..." });
 * const agent = phone.agent({
 *   engine: new GeminiLive({ voice: "Puck", model: "gemini-3.1-flash-live-preview" }),
 *   systemPrompt: "You are a friendly receptionist.",
 *   firstMessage: "Hello! How can I help?",
 * });
 * ```
 */

/** Constructor options for the `GeminiLive` engine marker. */
export interface GeminiLiveOptions {
  /** API key. Falls back to GEMINI_API_KEY / GOOGLE_API_KEY env vars when omitted. */
  apiKey?: string;
  /** Gemini Live model. Defaults to the adapter's native-audio default when omitted. */
  model?: string;
  /** Prebuilt voice name. Defaults to `Puck`. */
  voice?: string;
  /** Language code. Defaults to `en-US`. */
  language?: string;
  /** Sampling temperature. Defaults to `0.8`. */
  temperature?: number;
  /**
   * Enable native-audio affective dialog — the model adapts its tone/prosody to
   * the caller's emotion. Opt-in (default off) for backward compatibility.
   */
  affectiveDialog?: boolean;
  /**
   * Enable native-audio proactive audio — the model decides when to respond and
   * can stay silent to non-directed speech. Opt-in (default off).
   */
  proactiveAudio?: boolean;
  /**
   * Tune the server-side voice-activity detector. `startSensitivity: 'LOW'`
   * rejects background noise/breathing; `endSensitivity: 'HIGH'` + a lower
   * `silenceDurationMs` cut the wait before the model replies. Opt-in.
   */
  vad?: GeminiVadOptions;
  /**
   * Enable the model's "thinking" (chain-of-thought) on this voice session.
   *
   * `false` (DEFAULT) — thinking is disabled (`thinkingBudget: 0`). A voice
   * agent must never speak or transcribe its private reasoning; on the
   * Gemini 3.x native-audio models the chain-of-thought otherwise leaks into
   * BOTH the spoken audio and the transcript (e.g. "**Initiating Call
   * Protocol**…" before the real reply). Keep this off for telephony.
   *
   * `true` — re-enables dynamic thinking (the model decides how much to
   * think). Only do this for non-voice / latency-tolerant flows.
   */
  thinking?: boolean;
  /**
   * Explicit Gemini thinking budget (tokens). Takes precedence over
   * {@link thinking}. `0` disables thinking, `-1` lets the model decide
   * (dynamic). Omit to use the {@link thinking}-derived default (0 = off).
   */
  thinkingBudget?: number;
  /**
   * Gemini API version for the Live channel. Auto-detected from the model
   * name when omitted: native-audio output models — including
   * `gemini-3.1-flash-live-preview` — are served on `v1alpha`; all others use
   * the SDK default (`v1beta`). Pass `'v1alpha'` / `'v1beta'` to override when
   * a model is unavailable on the auto-detected channel.
   */
  apiVersion?: string;
}

/** Server-side VAD tuning for the Gemini Live native-audio engine. */
export interface GeminiVadOptions {
  /** 'LOW' = less sensitive (ignores noise); 'HIGH' = triggers on quieter sound. */
  startSensitivity?: 'HIGH' | 'LOW';
  /** 'HIGH' = detects end-of-turn faster (lower latency); 'LOW' = waits longer. */
  endSensitivity?: 'HIGH' | 'LOW';
  /** Silence (ms) before the turn is considered over. Lower = snappier replies. */
  silenceDurationMs?: number;
  /** Audio (ms) retained before detected speech start. */
  prefixPaddingMs?: number;
}

/**
 * Gemini Live engine marker — selects Google's native-audio Live API.
 */
export class GeminiLive {
  readonly kind = "gemini_live" as const;
  readonly apiKey: string;
  readonly model?: string;
  readonly voice?: string;
  readonly language?: string;
  readonly temperature?: number;
  readonly affectiveDialog?: boolean;
  readonly proactiveAudio?: boolean;
  readonly vad?: GeminiVadOptions;
  readonly thinking?: boolean;
  readonly thinkingBudget?: number;
  readonly apiVersion?: string;

  constructor(opts: GeminiLiveOptions = {}) {
    const key =
      opts.apiKey ?? process.env.GEMINI_API_KEY ?? process.env.GOOGLE_API_KEY;
    if (!key) {
      throw new Error(
        "Gemini Live requires an apiKey. Pass { apiKey: '...' } or set " +
          "GEMINI_API_KEY (or GOOGLE_API_KEY) in the environment.",
      );
    }
    this.apiKey = key;
    this.model = opts.model;
    this.voice = opts.voice;
    this.language = opts.language;
    this.temperature = opts.temperature;
    this.affectiveDialog = opts.affectiveDialog;
    this.proactiveAudio = opts.proactiveAudio;
    this.vad = opts.vad;
    this.thinking = opts.thinking;
    this.thinkingBudget = opts.thinkingBudget;
    this.apiVersion = opts.apiVersion;
  }
}

File: libraries/python/getpatter/engines/gemini.py
"""Gemini Live engine marker for Patter.

Selects Google's Gemini Live native-audio API (audio-in -> audio-out). The
client dispatches to
:class:`getpatter.providers.gemini_live.GeminiLiveAdapter` (through the
carrier-codec shim in :mod:`getpatter.providers.gemini_live_bridge`) when this
marker is passed to ``Patter.agent(engine=...)``.

Like the other engine markers this is a tiny immutable config object: it carries
credentials / voice / model / audio rates only. The runtime session lives in the
adapter, built server-side at call time.

Options present on the TypeScript ``GeminiLive`` marker but intentionally
ABSENT here, because the Python ``GeminiLiveAdapter`` has no constructor
argument that could honour them (forwarding them would be a silent no-op):

``affectiveDialog``, ``proactiveAudio``, ``vad``, ``thinking``,
``thinkingBudget``
    Native-audio / VAD / chain-of-thought knobs the TS adapter writes into its
    Live ``config``. The Python adapter builds its config without them.
``apiVersion``
    The Python adapter pins ``v1alpha`` (native-audio models are v1alpha-only)
    and exposes no override.

Add the matching adapter arguments first if these are needed; do not widen this
marker on its own.
"""

from __future__ import annotations

import os
from dataclasses import dataclass, field

from getpatter.providers.gemini_live import (
    DEFAULT_INPUT_SAMPLE_RATE_HZ,
    DEFAULT_OUTPUT_SAMPLE_RATE_HZ,
    GeminiLiveModel,
    GeminiLiveVoice,
)

__all__ = ["GeminiLive", "GEMINI_API_KEY_ENV_VARS"]

# Env vars checked, in order, when ``api_key`` is left empty. Mirrors
# ``getpatter.llm.google.LLM``: the Gemini-specific name wins over the broader
# Google one. Same pair (and order) as the TypeScript marker.
GEMINI_API_KEY_ENV_VARS: tuple[str, ...] = ("GEMINI_API_KEY", "GOOGLE_API_KEY")

# Patter's default Gemini Live model. Newer than the adapter's own default
# (which stays put for direct ``GeminiLiveAdapter`` users) and identical to the
# TypeScript engine default.
DEFAULT_MODEL = GeminiLiveModel.LIVE_3_1_FLASH_PREVIEW.value
DEFAULT_VOICE = GeminiLiveVoice.PUCK.value


@dataclass(frozen=True)
class GeminiLive:
    """Gemini Live engine config — selects Google's native-audio Live API.

    Holds the settings the Patter server needs to instantiate
    :class:`getpatter.providers.gemini_live.GeminiLiveAdapter` at call time.
    All fields are optional with safe defaults; knobs left at ``None`` are
    omitted so the adapter's own defaults stay authoritative.

    Example::

        from getpatter import Patter, Twilio, GeminiLive

        phone = Patter(carrier=Twilio(), phone_number="+1...")
        agent = phone.agent(
            engine=GeminiLive(voice="Puck"),
            system_prompt="You are a friendly receptionist.",
            first_message="Hello! How can I help?",
        )
    """

    # ``repr=False``: the marker is routinely printed in READMEs, docstrings
    # and error paths, and a default dataclass repr would put the raw key on
    # stdout / in a log line.
    api_key: str = field(default="", repr=False)
    model: str = DEFAULT_MODEL
    voice: str = DEFAULT_VOICE
    # BCP-47 language for the spoken reply. ``None`` defers to
    # ``Patter.agent(language=...)`` and then the adapter default (``en-US``).
    language: str | None = None
    # Sampling temperature. ``None`` keeps the adapter default (0.8).
    temperature: float | None = None
    # PCM16 mono rates Gemini negotiates on the Live channel. The carrier shim
    # resamples the 8 kHz telephony leg to/from these, so changing them changes
    # the resampling chain, not the wire format the carrier sees.
    input_sample_rate: int = DEFAULT_INPUT_SAMPLE_RATE_HZ
    output_sample_rate: int = DEFAULT_OUTPUT_SAMPLE_RATE_HZ

    def __post_init__(self) -> None:
        # Unlike the other engine markers this does NOT raise on a missing key:
        # ``GeminiLive()`` must stay constructible for introspection / docs on
        # a machine with no Google credentials. The call path fails fast
        # instead — ``Patter.agent()`` rejects a keyless Gemini Live agent at
        # build time, long before a call is dialled.
        if self.api_key:
            return
        for env_var in GEMINI_API_KEY_ENV_VARS:
            key = os.environ.get(env_var, "")
            if key:
                object.__setattr__(self, "api_key", key)
                return

    @property
    def kind(self) -> str:
        """Stable discriminator used for engine dispatch."""
        return "gemini_live"
